src/System.Linq/src/System/Linq/SequenceEqual.cs

```diff
@@ -28,9 +28,25 @@ public static bool SequenceEqual<TSource>(this IEnumerable<TSource> first, IEnum
                 throw Error.ArgumentNull(nameof(second));
             }
 
-            if (first is ICollection<TSource> firstCol && second is ICollection<TSource> secondCol && firstCol.Count != secondCol.Count)
+            if (first is ICollection<TSource> firstCol && second is ICollection<TSource> secondCol)
             {
-                return false;
+                if (firstCol.Count != secondCol.Count)
+                {
+                    return false;
+                }
+
+                if (firstCol is IReadOnlyList<TSource> firstList && secondCol is IReadOnlyList<TSource> secondList)
```

Comment:
So I did understand your comment correctly... I might be missing something here but looking at the code for ImmutableList, it would appear that it does implement IList? https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs

Also, we need the generic version in order to access typed elements in the loop for our comparer. What we could do is limit this to a TSource[] which, however, would limit the use of that change badly and slow down any other collection...